apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -152,6 +247,24 @@ type Patch struct {
 	// input to be transformed.
 	// +optional
 	Transforms []Transform `json:"transforms,omitempty"`
+
+	// Type can be set to refer to a named patch defined against the parent
+	// composition. If a Patch has a type set, then all other fields will be
+	// ignored and the referenced patch will be used instead.
+	// +optional
+	Type PatchType `json:"type,omitempty"`
```

Comment:
Let's say we have only the `Type` field. The logic users would think that there is a global pool of patches that I can choose from and I can contribute to that pool via `spec.patchSets`. Instead, we have two layers of classification; type of the patch, which could be patchSet or something else and also ID of a patch in that pool. To me, it feels like having the global pool is a simpler model for users to tinker when they write their compositions because number of custom compositions will almost never reach 10. So, having two layer identification pools feels kind of unnecessary.

We can have only `Type` field (or `Name`), some hard-coded types like `FromComposite`, `FromHTTPServer` etc. and the ones that users write in `spec.patchSets` (or preferably `spec.customPatchTypes` if you go with `Type`) like `MyRegionPatch`. Then when user chooses one, we'd first go over the hard-coded ones and if we can't find it, we'll iterate over the ones they supplied.

Reply:
This is the approach I tried to take first and failed, so went the simpler route to start with 🙃 - partly as I've never worked with the Kubernetes CRD spec before and confused myself with recursive patch types.

I think I have a good handle on the implementation for this now, but it means that when defining a patch in `spec.customPatchTypes`, we either need to define a `Type` and a `Name` (since the base type of the custom patch needs to be known), or we need to restrict the patches defined under `spec.customPatchTypes` to be a single `Patch` type - in this case I would use the `PatchSet` type as other patch types can be nested within it.


Assume we *use* the predefined and custom patches like this:

```yaml
  resources:
  - base:
      # ...
    patches:
    - type: Metadata # Dereferenced
    - type: PatchSet # Inlined PatchSet, handled as-is (pointless but here as an example)
      patches:
      - fromFieldPath: spec.parameters.namePrefix
        toFieldPath: metadata.generateName
    - fromFieldPath: spec.parameters.blah # Inlined FromCompositeFieldPath
      toFieldPath: spec.forProvider.providerField
    - type: Parameters # Dereferenced
```

We have two custom patch types - `Metadata` and `Parameters`.

If we allow type definition in `customPatchTypes` they can be defined like this:

```yaml
apiVersion: apiextensions.crossplane.io/v1beta1
kind: Composition
spec:
  customPatchTypes:
  # Custom PatchSet called Metadata
  - name: Metadata # Dereferenced as PatchSet
    type: PatchSet
    patches:
    - type: FromCompositeFieldPath
      fromFieldPath: metadata.labels
    - fromFieldPath: metadata.annotations # Type defaulted to FromCompositeFieldPath

  # Custom FromCompositeFieldPath called Parameters
  - name: Parameters # Dereferenced as individual Patch
    type: FromCompositeFieldPath
    fromFieldPath: spec.parameters.location
    toFieldPath: spec.forProvider.locationConstraint
```

Where `Metadata` is dereferenced to a `PatchSet` and `Parameters` is dereferenced to a single `FromCompositeFieldPath` Patch.

If we restrict `spec.customPatchTypes` to only defining `PatchSet` instances:

```yaml
apiVersion: apiextensions.crossplane.io/v1beta1
kind: Composition
spec:
  customPatchTypes:
  # Custom PatchSet called Metadata
  - type: Metadata # Dereferenced as PatchSet
    patches:
    - type: FromCompositeFieldPath
      fromFieldPath: metadata.labels
    - fromFieldPath: metadata.annotations # Type defaulted to FromCompositeFieldPath

  # Custom FromCompositeFieldPath nested inside a PatchSet
  - type: Parameters # Dereferenced as PatchSet
    patches:
    - fromFieldPath: spec.parameters.location
      toFieldPath: spec.forProvider.locationConstraint
```

In practice the differences are small - but I _think_ if we restrict `spec.customPatchTypes` to being **only** `PatchSet` instances, then we avoid the need of the additional name field, we get consistency between the `type` field on `resource` patches as well as `spec.customPatchTypes` and we have just as much flexibility, it just means if you want to use a different patch type in `spec.customPatchTypes` you have to nest it in the `patches` field.

Does this make sense? 

From an implementation perspective, `PatchSet` just defines a list of `patches` and an `Apply` method which loops the `patches` and runs their `Apply` method.

Reply:
The `customPatchSets` pattern feels like a regression to me compared to the `patchSets` API where the type of patch the and the identifier/name were distinct.

It seems to me that we are overloading the `type` field to cover both the "built in" types of patch and the extended types of patch (from the `customPatchSets` array). As far as I can tell the main advantage to this approach is needing one fewer field (i.e. `type` does double duty and covers the `name` or `id` field). The main disadvantage is that it makes it harder to distinguish what the type field is referring to at a glance. Is `type: Widget` a native type or defined by `customPatchTypes`? What would a user new to composition think if four of of five patch types were declared in the `Composition`, but the fifth was some magic value they couldn't figure out the source of?

It's also a little odd to me that a custom "type of patch" is actually a custom _set of patches_.

Reply:
I kinda liked the succinctness of overloading the type field but you definitely have a solid point there - the `type` field either referred to a `PatchType` (of which `PatchSet` is one) but also an _instance_ of `PatchSet` with a particular name. I can definitely see that being confusing for new users when working out where patches are coming from.

Also, if the type field can only contain known values, then it can be validated using the `+kubebuilder:validation:Enum=` setting, which avoids a situation where you typo an inbuilt patch type but don't get an error until the composite resource is created, because that is the point where `PatchSet` items from the Composition are injected into the composed resource.

I've changed the API back to using `patchSets` since this is effectively what they were anyway, and no longer override the type field within the defined patches, but use the name instead.

Now when applying an invalid Composition I get: `The Composition "aws-object-storage-bucket" is invalid: spec.resources.patches.type: Unsupported value: "PatchxxXY": supported values: "FromCompositeFieldPath", "PatchSet"`

Reply:
> The main disadvantage is that it makes it harder to distinguish what the type field is referring to at a glance. Is type: Widget a native type or defined by customPatchTypes?

How much would they be concerned with knowing from where it comes from though? Even if they are, it's a top-level field. The non-author readers would see the custom patch types immediately as well since it's a top-level field.

> What would a user new to composition think if four of of five patch types were declared in the Composition, but the fifth was some magic value they couldn't figure out the source of?

The fact that the name of the field is `customPatchTypes` already implies that there is an existing set of patch types that author made addition to. So, I feel like it's not that much of a surprise.

> It's also a little odd to me that a custom "type of patch" is actually a custom set of patches.

I think I can see that being confusing. So, I'd like us examine a similar case in Kubernetes API where users first define the set of `volumes` and then use them in the `containers` array of the pod, under `volumeMounts`:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: configmap-pod
spec:
  containers:
    - name: test
      image: busybox
      volumeMounts:
        - name: config-vol
          mountPath: /etc/config
  volumes:
    - name: config-vol
      configMap:
        name: log-config
        items:
          - key: log_level
            path: log_level
```

IMO, `volumes` maps to the custom patches array and `volumeMounts` could map to `patches` in the `resources` array. So, instead of trying so hard to reuse `fromFieldPath` and `toFieldPath` for many things, we can have a type distinction in the field name and let each type has the diverse set of fields and mechanism available to them. The examples above would look like the following:
```yaml
apiVersion: apiextensions.crossplane.io/v1beta1
kind: Composition
spec:
  customPatches:
  - name: Metadata
    resourcePatches: # The mechanism of the patch is identified by the field used here.
    - fromFieldPath: metadata.labels
    - fromFieldPath: metadata.annotations
  - name: Location
    resourcePatch:
      fromFieldPath: spec.parameters.location
      toFieldPath: spec.forProvider.locationConstraint
resources:
- base:
   # ...
  patches:
  - name: Metadata # Name is given only if you want to use the custom patches you named.
  - name: Location
  # compositePatch means we're patching the composite.
  - compositePatch: # No name is necessary if you will use built-in patches. 
      fromFieldPath: status.atProvider.ARN
      toFieldPath: status.rdsARN
```

One difference between volumes semantics is that they don't have any built-in volumes so they always use `name` in the `containers[*].volumeMounts[*]`. In our case, user will give a name in `resources[*].patches[*]` if they'd like to use the custom patches they defined. If the patch is unique to that specific resource, then they won't need to use `name` field, they will just configure the patch to define the mechanism. This would also make it clear from where the patch comes from.

Reply:
@muvaf I think the example of volumes makes sense, but we're going to need patch 'types' as well as 'names' that are references into what you've called the `customPatches` array. So to me, this takes us back to the API in which a reference to the array of reusable patches (by name or ID) is a distinct type of patch. In general, this feels like a "more is more" situation. I can't see what we gain by overloading the type/name concept except for one fewer field for folks to type, at the expense of more cognitive load figuring out what that field does.

Reply:
> I can't see what we gain by overloading the type/name concept except for one fewer field for folks to type, at the expense of more cognitive load figuring out what that field does.

I'm leaning towards succinctness because I think custom patches is an advanced use case where we can expect some experience. And even in cases where you don't define your own custom patches, this allows each built-in type to have its own set of fields separately from others. For making it more clear, the following pattern where we explicitly reference a custom patch could be easier to grasp.

```yaml
apiVersion: apiextensions.crossplane.io/v1beta1
kind: Composition
spec:
  customPatches: # The struct here has 'name' field and inlined patch struct.
  - name: Metadata
    resourcePatches: # The mechanism of the patch is identified by the field used here.
    - fromFieldPath: metadata.labels
    - fromFieldPath: metadata.annotations
  - name: Location
    resourcePatch:
      fromFieldPath: spec.parameters.location
      toFieldPath: spec.forProvider.locationConstraint
resources:
- base:
   # ...
  patches:
  - customPatchRef: # One of patch types in the patch struct.
      name: Metadata
  - customPatchRef:
      name: Location
  - compositePatch: # compositePatch means we're patching the composite, it's a type of patch.
      fromFieldPath: status.atProvider.ARN
      toFieldPath: status.rdsARN
  - resourcePatchList: # resourcePatchList is a list of resource patches where composed is patched.
    - fromFieldPath: spec.size
      toFieldPath: spec.forProvider.dbSize
```

Anyway, I believe this API is more inline with Kubernetes API, but I can see that the suggested top-level name/type separation approach is also acceptable.

Reply:
@muvaf I responded to your similar comment in https://github.com/crossplane/crossplane/issues/1974. In short, it seems like your latest proposal is a breaking API change and would also benefit from a `type` discriminator field.

Reply:
Resolved in https://github.com/crossplane/crossplane/issues/1974